### PR TITLE
chore: new exportable indexer and datastore packages

### PR DESCRIPTION
**Pull request by crozzy**
**High Level Changes**:
* Unify all the postgres store implementations under `datastore/postgres`
* Move `/internal/indexer` to `/indexer`
* Add func to `store.AffectedManifests()` signature to allow filtering logic
* Change instantiation of `Libindex`:
  * Make FetchArena interface
  * Make Locker interface
  * Make Store interface configurable

**Not tackled**:
* Move `/internal/matcher` to `/matcher`
* Changing instantiation of `Libvuln`

**Example Usage**:
* https://github.com/crozzy/local-clair/blob/main/main.go

### CODE REVIEW

File: internal/datastore/postgres/affectedmanifest.go [hunk 1]
```diff
@@ -179,10 +178,9 @@ WHERE
 	// for each package discovered create an index record
 	// and determine if any in-tree matcher finds the record vulnerable
 	var filteredRecords []claircore.IndexRecord
-	om := omnimatcher.New(nil)
 	for _, pkg := range pkgsToFilter {
 		pr.Package = &pkg
-		match, err := om.Vulnerable(ctx, &pr, &v)
+		match, err := vulnFunc(ctx, &pr, &v)
```
Comment (crozzy) [inline internal/datastore/postgres/affectedmanifest.go:183]: This is an issue as either way I look at it we need to change the interface. It's fairly clean accepting a func but in theory there shouldn't be application logic here so maybe the function should be split and the caller does the filtering on their side.

File: docs/SUMMARY.md [hunk 2]
```diff
@@ -18,8 +18,8 @@
   - [Distribution Scanner](./reference/distribution_scanner.md)
   - [Ecosystem](./reference/ecosystem.md)
   - [Index Report](./reference/index_report.md)
-  - [Libindex Store](./reference/libindex_store.md)
-  - [Libvuln Store](./reference/libvuln_store.md)
+  - [Indexer Store](./reference/indexer_store.md)
+  - [Datastore MatcherStore](./reference/matcher_store.md)
```
Comment (hdonnay) [inline docs/SUMMARY.md:22]:
```suggestion
  - [Matcher Store](./reference/matcher_store.md)
```

File: docs/reference/matcher_store.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Datastore MatcherStore
+The `datastore.MatcherStore` interface defines all necessary persistence methods for `Libvuln`
+to provide its functionality.
+
+{{# godoc datastore.MatcherStore}}
```
Comment (hdonnay) [inline docs/reference/matcher_store.md:5]: Should this just be referred to as the "Matcher Store" in prose? "Datastore MatcherStore" makes my eye twitch.

File: vulnerability.go [hunk 4]
```diff
@@ -38,3 +39,5 @@ type Vulnerability struct {
 	// compared.
 	ArchOperation ArchOp `json:"arch_op,omitempty"`
 }
+
+type CheckVulnernableFunc func(ctx context.Context, record *IndexRecord, vuln *Vulnerability) (bool, error)
```
Comment (hdonnay) [inline vulnerability.go:43]: Needs some documentation.

**Review by hdonnay — CHANGES_REQUESTED** [review]

File: libindex/options.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	// scanners. They're broken out by kind, and only used if a scanner
+	// implements the appropriate interface.
+	//
+	// Providing a function for a scanner that's not expecting it is not a fatal
+	// error.
+	ScannerConfig struct {
+		Package, Dist, Repo map[string]func(interface{}) error
+	}
+	// vscnrs is a convenience method for holding a list of versioned scanners
+	vscnrs indexer.VersionedScanners
```
[the remark this answers is not in the document]
Comment (crozzy) [reply libindex/options.go:59]: Shouldn't be difficult to move `vscnrs` as a private member of `Libindex` if that's what you're thinking, it doesn't really make sense there as an "option" when it's just used by `Libindex` in the internal implementation.

File: libindex/init.go [hunk 6]
```diff
@@ -8,46 +8,16 @@ import (
...
-	}
-
-	return pool, nil
-}
-
-// initialize a indexer.Store given libindex.Opts
-func initStore(_ context.Context, pool *pgxpool.Pool, opts *Opts) (indexer.Store, error) {
-	cfg, err := pgx.ParseConfig(opts.ConnString)
+// InitPostgresStore initialize a indexer.Store given libindex.Opts
+func InitPostgresStore(_ context.Context, pool *pgxpool.Pool, doMigration bool) (indexer.Store, error) {
```
[the remark this answers is not in the document]
Comment (crozzy) [reply libindex/init.go:19]: In the matcher side I end up moving the migrations over to the datastore too so it makes sense to have it all here

[the remark this answers is not in the document]
Comment (crozzy) [reply libindex/init.go:19]: https://github.com/crozzy/claircore/tree/crozzy-matcher-refactor/datastore/postgres/migrations

[the remark this answers is not in the document]
Comment (crozzy) [reply libindex/init.go:19]: https://github.com/quay/claircore/commit/34a38fdce99c5d39e51a59ae44ebdb1c481e7a93 is this the change you're referring to?

[the remark this answers is not in the document]
Comment (hdonnay) [reply libindex/init.go]: no, looks like I made those on another branch, I'll pick them off. I'm thinking of https://github.com/quay/claircore/pull/607/commits/02f679c2a81495c69c9379ce4d8837b59a4b0a3a and https://github.com/quay/claircore/pull/607/commits/16f4abdaa8d403880dee4524a966ff0c2def6c91

**Comment by crozzy** [comment]
I rebased `restructure-dev` and then rebased this branch on top so we should be pretty up to date

File: test/mock/vulnstore/generate.go [hunk 7]
```diff
@@ -1,4 +0,0 @@
-package mock_vulnstore
```
[the remark this answers is not in the document]
Comment (crozzy) [reply test/mock/vulnstore/generate.go:1]: moved now, thanks :+1:

**Review by hdonnay — CHANGES_REQUESTED** [review]

File: datastore/postgres/init.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+	"fmt"
+
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/quay/claircore/pkg/poolstats"
+	"github.com/quay/zlog"
+)
+
+// InitDB initialize a postgres pgxpool.Pool based on the connection string
+func InitDB(ctx context.Context, connString string, applicationName string) (*pgxpool.Pool, error) {
```
[the remark this answers is not in the document]
Comment (crozzy) [reply datastore/postgres/init.go]: good point, renamed file and function accordingly, scans better when called

**Review by hdonnay — APPROVED** [review]

**Pull request merged**